SNIClient.py

```diff
@@ -511,7 +511,8 @@ async def snes_recv_loop(ctx: SNIContext) -> None:
             ctx.snes_autoreconnect_task = asyncio.create_task(snes_autoreconnect(ctx), name="snes auto-reconnect")
 
 
-async def snes_read(ctx: SNIContext, address: int, size: int) -> typing.Optional[bytes]:
+async def snes_read(ctx: SNIContext, address: int, size: int) -> bytes:
```

Comment:
I think this is not a good solution to this problem.

There aren't any good tools to make sure the exception is handled. <details><summary>...</summary>The lack of tools to make sure exceptions are handled isn't because of a lack of motivation or people to make the tools. Many have discussed it and tried, and have found there's no good way to make it work in Python.</details>

With the `Optional[bytes]` return type, there are good tools to make sure the failed read is handled.

It is a problem that many people don't use those tools to make sure the `None` return value is handled, but that doesn't mean we should take that avenue away from people that do use those tools.

We can't force people to use the tools that are available.

So we need to do better at handling any bad code from `SNIClient` implementations.

`validate_rom` can be run from any installed apworld, whether we're playing that game or not. So I think we should catch any `Exception` thrown (not just some specific exception) and log it and then treat it as if it returned `False`.

`game_watcher` should only be run if we're actually playing that game, so it can be less forgiving.
If the player isn't seeing any error message, then we should change that to make it more visible, so they can report the problem to the world developer and get the problem fixed.

Reply:
`game_watcher` is exactly where we need to be more resilient though, as 99% of SNIClient failures occur there. Ultimately, while we want people to properly check the result from `snes_read`, we cannot assume they will do so. 

At least to me, there exists two possible solutions. We catch every exception resulting from `validate_rom`, show an error, and continue, or do as this PR does and let any failed SNI reads fall through while preserving current behavior for any real exceptions. Either way, the exception *must* be handled in some manner. Not doing so and letting the thread die causes major degradation in the usability of the client (see all of the SNIClient issues posted in the discord, about 95% of them can be traced back to this singular issue).

Reply:
I would be ok with handling the exceptions and letting the `game_watcher` task continue, so the player might be able to keep playing, but it's not ok to assume that that means there's no problem. A `snes_read` raising an exception might leave the context in an invalid state. And removing the `None` from the return type of `snes_read` disables tools that can help developers make sure they're keeping the context in a valid state.

We should tell the user that an error happened so the bug can be reported and handled properly by the world dev.

If we just assume that a `snes_read` failure is ok, that will make finding and fixing bugs more difficult.

Reply:
Just wanted to say that we do have ways to make sure that the exception is handled in this case:

Since we have control of the calling context in `game_watcher(ctx: SNIContext)`, we can have something custom happen at *runtime* there in the `except` block.

Imo, the way this should work is that people are expected to handle the Exception, but if they don't, it will be caught by *our* code instead, and then something happens, such as an annoying message box for example.

Reply:
> Just wanted to say that we do have ways to make sure that the exception is handled in this case:

It seems that this is trying to reply to this (correct me if I'm wrong).

> There aren't any good tools to make sure the exception is handled.

But in that specific context (different from other contexts in the same discussion), "handling the exception" means ensuring the state of the application is good.
We cannot do that.
Handling an exception well requires examining all of the code that the exception goes through. That means examining apworlds that haven't been written yet. Only the world developer can handle exceptions well.

> people are expected to handle the Exception

What I'm pointing out is that there are no tools to help remind the developer to handle exceptions like there are with an error return value.

We could say "people are expected to handle the Exception", but we can already see that that isn't going to happen.

There are good reasons that newer languages like rust, go, and zig have no exceptions, and instead put errors in the return values of the function.
There's no way to handle exceptions well in a scalable, maintainable way. (Exceptions can be handled well in small toy examples that never need to be maintained.)
Putting the error in the return value is the better design. (It doesn't have to be `None` - but as I pointed out in the other PR, that's not really relevant to the problem this PR is trying to address.) What's relevant to this PR is that it's moving things in a bad direction by changing from a return value error to an exception error.

Reply:
But yes, I also like the idea of making a custom "error state" class that fails on equality check with bool or anything else that ppl might do with the None that they shouldn't

Reply:
> Correct, but we can at least encourage them to do so, by making them either
> 
>     1. actively write bad-looking code to avoid it that will stare them in the face every time
> 
>     2. Have something annoying-but-not-game-breaking for the user that will make users ask the dev for an update

1. I don't understand what "bad-looking code" you're imagining they will have to write (and how we couldn't do the same with a return value error).
2. This is the same whether it's an exception or return value error. (And the return value error has the further advantages I've already pointed out.)

Reply:
Wrapping their whole game watcher function in a try-except

Reply:
But yeah I'm fine with the return value error

Reply:
> Wrapping their whole game watcher function in their own singular try-except

Ah, so what KDL3 is already doing in order to avoid this issue.